cpp/merkletree/sparse_merkle_tree.cc

```diff
@@ -0,0 +1,216 @@
+#include "cpp/merkletree/sparse_merkle_tree.h"
+
+#include <stddef.h>
+#include <algorithm>
+#include <vector>
+
+#include "merkletree/merkle_tree_math.h"
+#include "util/util.h"
+
+using std::make_pair;
+using std::ostream;
+using std::ostringstream;
+using std::reverse;
+using std::string;
+using std::unique_ptr;
+using std::unordered_map;
+using std::vector;
+
+
+const vector<string>* GetNullHashes(const TreeHasher& hasher) {
+  static unique_ptr<const vector<string>> null_hashes;
+  if (!null_hashes) {
+    vector<string> r{hasher.HashLeaf("")};
+    for (int i(1); i < hasher.DigestSize() * 8; ++i) {
+      r.emplace_back(hasher.HashChildren(r.back(), r.back()));
+    }
+    reverse(r.begin(), r.end());
+    null_hashes.reset(new vector<string>(std::move(r)));
+  }
+  return null_hashes.get();
+}
+
+
+SparseMerkleTree::SparseMerkleTree(SerialHasher* hasher)
+    : serial_hasher_(CHECK_NOTNULL(hasher)->Create()),
+      treehasher_(hasher),
+      null_hashes_(GetNullHashes(treehasher_)),
+      root_hash_(null_hashes_->at(0)) {
+}
+
+
+void SparseMerkleTree::EnsureHaveLevel(size_t level) {
+  if (tree_.size() < (level + 1)) {
+    tree_.resize(level + 1);
+  }
+}
+
+
+void SparseMerkleTree::SetLeafByHash(const string& key, const string& data) {
+  serial_hasher_->Reset();
+  serial_hasher_->Update(key);
+  return SetLeaf(PathFromBytes(serial_hasher_->Final()), data);
+}
+
+
+void SparseMerkleTree::SetLeaf(const Path& path, const string& data) {
+  CHECK_EQ(treehasher_.DigestSize(), path.size());
+  // Mark the tree dirty:
+  root_hash_.clear();
+
+  IndexType node_index(0);
+  for (int depth(0); depth <= kDigestSizeBits; ++depth) {
```

Comment:
Don't want to do that - there's a big WARNING comment about how 64 is less that 256 in the header somewhere, I'll be replacing the `uin64_t` which `IndexType` currently is defined as with a `BigNum` wrapper RSN.